can_validator/templates/special_cases.txt

```diff
@@ -0,0 +1,223 @@
+  type_out->current_limited_via_igbt_temp = BIT_GET(data_string, 23);
+  type_out->current_limited_to_continuous_via_igbt_temp = BIT_GET(data_string, 24);
+  type_out->current_reduction_low_frequency = BIT_GET(data_string, 25);
+  type_out->current_reduction_via_motor_temp = BIT_GET(data_string, 26);
+  type_out->current_reduction_via_analog_input = BIT_GET(data_string, 27);
+  type_out->current_accumulator_limit_charged = BIT_GET(data_string, 28);
+  type_out->handwheel_input_selected = BIT_GET(data_string, 31);
+}
+
+TO_CAN(Can_MC_DataReading) {
+  uint64_t bitstring = 0;
+  bitstring = INSERT(type_in->type, bitstring, 0, 8);
+
+  // Little Endian
+  int16_t swap_value = swap_int16(type_in->value);
+
+  bitstring = INSERT(swap_value, bitstring, 8, 16);
+  from_bitstring(&bitstring, can_out->data);
+  can_out->id = MC_RESPONSE__id;
+  can_out->len = 3;
+}
+
+FROM_CAN(Can_MC_DataReading) {
+  uint64_t bitstring = 0;
+  to_bitstring(can_in->data, &bitstring);
+  type_out->type = (Can_MC_RegID_T)(EXTRACT(bitstring, 0, 8));
+
+  // Little Endian
+  int16_t swap_value = (int16_t)(swap_uint16(EXTRACT(bitstring, 8, 16)));
+
+  type_out->value = swap_value;
+}
+
+TO_CAN(Can_Vcu_MCTorque) {
+  uint64_t bitstring = 0;
+  bitstring = INSERT(CAN_MC_REG_TORQUE_CMD, bitstring, 0, 8);
+  // TODO all other devices are big endian except motor controllers,
+  // so we want to reverse the endianness
+  uint64_t temp_torque_cmd = type_in->torque_cmd;
+  uint8_t high_byte = EXTRACT(temp_torque_cmd, 48, 8);
+  uint8_t low_byte = EXTRACT(temp_torque_cmd, 56, 8);
+  bitstring = INSERT(low_byte, bitstring, 8, 8);
+  bitstring = INSERT(high_byte, bitstring, 16, 8);
+  from_bitstring(&bitstring, can_out->data);
+  can_out->id = VCU_MC_MESSAGE__id;
+  can_out->len = 3;
+}
+
+FROM_CAN(Can_Vcu_MCTorque) {
+  uint64_t bitstring = 0;
+  to_bitstring(can_in->data, &bitstring);
+  // TODO all other devices are big endian except motor controllers,
+  // so we want to reverse the endianness
+  uint8_t low_byte = EXTRACT(bitstring, 8, 8);
+  uint8_t high_byte = EXTRACT(bitstring, 16, 8);
+  int16_t low_word = (int16_t) low_byte;
+  int16_t high_word = (int16_t) (high_byte << 8);
+  type_out->torque_cmd = low_word | high_word;
+}
+
+TO_CAN(Can_Vcu_MCRequest) {
+  uint64_t bitstring = 0;
+  bitstring = INSERT(CAN_MC_REG_MSG_REQUEST, bitstring, 0, 8);
+  bitstring = INSERT(type_in->requestType, bitstring, 8, 8);
+  bitstring = INSERT(type_in->period, bitstring, 16, 8);
+  from_bitstring(&bitstring, can_out->data);
+  can_out->id = VCU_MC_MESSAGE__id;
+  can_out->len = 3;
+}
+
+FROM_CAN(Can_Vcu_MCRequest) {
+  uint64_t bitstring = 0;
+  to_bitstring(can_in->data, &bitstring);
+  type_out->requestType = (Can_MC_RegID_T) (EXTRACT(bitstring, 8, 8));
+  type_out->period = EXTRACT(bitstring, 16, 8);
+}
+
+DEFINE(Can_Vcu_MCRequest)
+DEFINE(Can_Vcu_MCTorque)
+DEFINE(Can_MC_DataReading)
+DEFINE(Can_MC_ErrorAndWarning)
+DEFINE(Can_MC_State)
```

Comment:
This could also be a `.c` as with any other file that contains code, even partial.